### PR TITLE
NP-1204: Add type attribute to the JSON serializations of the DTO objects

### PR DESCRIPTION
**Pull request by axthosarouris**

### CODE REVIEW

File: src/main/java/no/unit/nva/database/UserDb.java [hunk 1]
```diff
@@ -18,7 +17,6 @@
 import nva.commons.utils.JacocoGenerated;
 
 @DynamoDBTable(tableName = "OverridenByEnvironmentVariable")
-@DynamoDBTyped
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/database/UserDb.java:21]: not necessary.

File: src/main/java/no/unit/nva/model/RoleDto.java [hunk 2]
```diff
@@ -12,7 +13,8 @@
 import nva.commons.utils.JsonUtils;
 import nva.commons.utils.StringUtils;
 
-public class RoleDto implements WithCopy<RoleDto.Builder>, JsonSerializable, Validable {
+@JsonTypeName("Role")
+public class RoleDto implements WithCopy<RoleDto.Builder>, JsonSerializable, Validable, Typed {
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/model/RoleDto.java:17]: interface `Typed` contains all the relevant Jackson annotations.

File: src/main/java/no/unit/nva/model/Typed.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
+package no.unit.nva.model;
+
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import com.fasterxml.jackson.annotation.JsonTypeInfo.As;
+import com.fasterxml.jackson.annotation.JsonTypeInfo.Id;
+
+@JsonTypeInfo(use = Id.NAME, include = As.PROPERTY, property = "type")
+@JsonSubTypes({
```
Comment (axthosarouris) [inline src/main/java/no/unit/nva/model/Typed.java:9]: Db objects do not need to be annotated as the JSON serialization is handled by the DynamoDb Mapper

File: src/test/java/features/ScenarioContext.java [hunk 4]
```diff
@@ -40,7 +40,7 @@ protected void setDatabaseService(DatabaseServiceImpl databaseService) {
 
     protected <T> GatewayResponse<T> getApiGatewayResponse(Class<T> responseBodyClass) throws IOException {
         JavaType typeRef = JsonUtils.objectMapper.getTypeFactory()
-            .constructParametrizedType(GatewayResponse.class, GatewayResponse.class, responseBodyClass);
+            .constructParametricType(GatewayResponse.class, responseBodyClass);
```
Comment (axthosarouris) [inline src/test/java/features/ScenarioContext.java]: `constructParamertrizedType` is obsolete. Using the parametricType instead.

File: src/test/resources/features/Users.feature [hunk 5]
```diff
@@ -24,7 +25,7 @@ Feature: Users
     And the authorized client forms a "GET" request
     And the request has the following path parameters:
       | parameter | value               |
-      | user      | someone@institution |
+      | username  | someone@institution |
```
Comment (axthosarouris) [inline src/test/resources/features/Users.feature:28]: will change it to userId in subsequent PR

**Review by janbj — DISMISSED** [review]

File: src/main/java/no/unit/nva/model/RoleDto.java [hunk 6]
```diff
@@ -12,7 +13,8 @@
 import nva.commons.utils.JsonUtils;
 import nva.commons.utils.StringUtils;
 
-public class RoleDto implements WithCopy<RoleDto.Builder>, JsonSerializable, Validable {
+@JsonTypeName("Role")
```
Comment (brinxmat) [inline src/main/java/no/unit/nva/model/RoleDto.java]: Can you rename the intefaces->interfaces package?

File: src/test/java/no/unit/nva/model/DtoTest.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+
+// Not a final class in order to show where these values methods are used through inheritance.
+@SuppressWarnings("NonFinalUtilityClass")
+public class DtoTest {
+
+    protected static final String JSON_TYPE_ATTRIBUTE = "type";
+    protected static final String USER_TYPE_LITERAL = "User";
+    protected static final String ROLE_TYPE_LITERAL = "Role";
```
Comment (brinxmat) [inline src/test/java/no/unit/nva/model/DtoTest.java]: USER_TYPE_LITERAL and ROLE_TYPE_LITERAL should be moved to the classes where they are used.

File: src/test/java/no/unit/nva/model/RoleDtoTest.java [hunk 8]
```diff
@@ -69,4 +79,43 @@ public void isValidReturnsFalseWhenUsernameIsNullOrBlank(String emptyOrNullRoleN
         setter.invoke(roleDto, emptyOrNullRoleName);
         assertThat(roleDto.isValid(), is(equalTo(false)));
     }
+
+    @DisplayName("RoleDto object contains type with value \"Role\"")
```
Comment (brinxmat) [inline src/test/java/no/unit/nva/model/RoleDtoTest.java:84]: Hmm. We need to discuss testing strategy

**Review by brinxmat — CHANGES_REQUESTED** [review]

File: src/test/java/no/unit/nva/model/RoleDtoTest.java (hunk 8 above)
Comment (janbj) [reply src/test/java/no/unit/nva/model/RoleDtoTest.java:84]: You can try converting it back to POJO instead.

**Review by janbj — DISMISSED** [review]

**Comment by axthosarouris** [comment]
I added tests for each handler

File: src/test/java/no/unit/nva/handlers/AddRoleHandlerTest.java [hunk 9]
```diff
@@ -134,6 +134,27 @@ public void errorMessageIsLoggedWhenAddRoleHandlerThrowsDatasyncException()
...
+        throws InvalidEntryInternalException, IOException {
+        RoleDto roleDto = createSampleRole();
+        ObjectNode objectWithoutType = createInputObjectWithoutType(roleDto);
+
+        InputStream requestInput = createRequestInputStream(objectWithoutType);
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+
+        addRoleHandler.handleRequest(requestInput, outputStream, context);
+        GatewayResponse<Problem> response = GatewayResponse.fromOutputStream(outputStream);
+        assertThat(response.getStatusCode(), is(equalTo(HttpStatus.SC_BAD_REQUEST)));
```
Comment (brinxmat) [inline src/test/java/no/unit/nva/handlers/AddRoleHandlerTest.java:145]: You can probably extract lines 140-147 to a method a la generateBadRequestRoleHasNoType

File: src/test/java/no/unit/nva/handlers/AddRoleHandlerTest.java [hunk 10]
```diff
@@ -134,6 +134,27 @@ public void errorMessageIsLoggedWhenAddRoleHandlerThrowsDatasyncException()
...
+
+        Problem problem = response.getBodyObject(Problem.class);
+        assertThat(problem.getDetail(), is(equalTo(InvalidOrMissingTypeException.MESSAGE)));
```
Comment (brinxmat) [inline src/test/java/no/unit/nva/handlers/AddRoleHandlerTest.java:148]: Given the above refactoring, you can split the two asserts into separate tests calling the extracted method

File: src/test/java/no/unit/nva/handlers/AddUserTest.java [hunk 11]
```diff
@@ -141,6 +143,24 @@ public void processInputThrowsDataSyncExceptionWhenDatabaseServiceCannotReturnSa
         assertThat(exception.getMessage(), containsString(SYNC_ERROR_MESSAGE));
     }
 
+    @DisplayName("handleRequest() returns BadRequest when input object has no type")
+    @Test
+    public void handlerRequestReturnsBadRequestWhenInputObjectHasNoType()
+        throws InvalidEntryInternalException, IOException {
+        UserDto sampleUser = createUserWithRolesAndInstitution();
```
Comment (brinxmat) [inline src/test/java/no/unit/nva/handlers/AddUserTest.java:150]: Similarly, I'd refactor this to move the setup to a named method

File: src/test/java/no/unit/nva/handlers/GetRoleHandlerTest.java [hunk 12]
```diff
@@ -49,6 +60,20 @@ public void statusCodeReturnsOkWhenRequestIsSuccessful() {
         assertThat(successCode, is(equalTo(HttpStatus.SC_OK)));
     }
 
+    @DisplayName("handleRequest returns Role object with type \"Role\"")
+    @Test
+    public void handleRequestReturnsRoleObjectWithTypeRole()
+        throws ConflictException, InvalidEntryInternalException, InvalidInputException, IOException {
+        addSampleRoleToDatabase();
```
Comment (brinxmat) [inline src/test/java/no/unit/nva/handlers/GetRoleHandlerTest.java:68]: Again, I'd probably move the set up to a named method

File: src/test/java/no/unit/nva/handlers/GetUserHandlerTest.java [hunk 13]
```diff
@@ -44,6 +54,27 @@ public void init() {
         context = mock(Context.class);
     }
 
+    @DisplayName("handleRequest returns User object with type \"User\"")
+    @Test
+    public void handleRequestReturnsRoleObjectWithTypeRole()
+        throws ConflictException, InvalidEntryInternalException, InvalidInputException, IOException {
+        requestInfo = createRequestInfoForGetUser(SOME_USERNAME);
```
Comment (brinxmat) [inline src/test/java/no/unit/nva/handlers/GetUserHandlerTest.java]: Again, move the setup to a named method

File: src/test/java/no/unit/nva/handlers/UpdateUserHandlerTest.java [hunk 14]
```diff
@@ -161,10 +163,29 @@ public void processInputReturnsNotFoundWhenHandlerWhenTryingToUpdateNonExistingU
         assertThat(problem.getDetail(), containsString(USER_NOT_FOUND_MESSAGE));
     }
 
-    private <T> GatewayResponse<T> sendUpdateRequest(String userId, UserDto userUpdate)
+    @Test
+    public void handleRequestReturnsBadRequestWhenInputRoleHasNoType()
+        throws InvalidEntryInternalException, IOException {
+        UserDto userDto = createSampleUser();
```
Comment (brinxmat) [inline src/test/java/no/unit/nva/handlers/UpdateUserHandlerTest.java]: Again

**Review by brinxmat — DISMISSED** [review]

**Review by brinxmat — APPROVED** [review]

**Pull request merged**